### PR TITLE
Fix hidden checkpoint parameter handling in local CUDA graphs

### PR DESCRIPTION
**Pull request by JF-D**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

Fixes local CUDA-graph training when `CheckpointWithoutOutput` recomputation uses parameters from nested modules that are not visible in the outer autograd graph.

## Problem                                                                                                                                               

  `CheckpointWithoutOutput` runs its recompute function during backward. Parameters
  used only by that function do not have autograd edges from the outer forward
  outputs.

  CUDA-graph capture previously discovered `params_to_backprop` only by traversing
  the autograd graph from the captured forward outputs. Consequently, a parameter
  used exclusively by checkpoint recomputation could be omitted from the captured
  backward inputs.
 
  This caused two related problems:

  1. The nested recompute backward accumulated the parameter gradient into
     `param.grad`, but `torch.autograd.grad` returned `None` for the corresponding
     outer parameter input. The gradient was therefore not transferred to
     `param.main_grad`.

  2. Because the parameter was not passed as a synthetic input to the CUDA-graph
     replay node, replay had no autograd edge to its `AccumulateGrad` node. The
     normal DDP hook registered on that node did not fire, leaving the DDP bucket
     with an incomplete ready count.

  This can result in missing gradients, incorrect gradient accumulation, or DDP
  waiting for a bucket that never becomes ready.

  Example:
  ```
  def forward(self, x):
      checkpoint = CheckpointWithoutOutput()


      hidden = checkpoint.checkpoint(self.checkpointed, x)
      output = self.projection(hidden)

      checkpoint.discard_output_and_register_recompute(output)
      return output
  ```
  `self.projection.weight` is used by the outer forward autograd graph, so it can be 
    discovered by traversing backward from output. In contrast, `self.checkpointed.weight`
    is closed over by the checkpoint recompute function. Its computation is deferred until
    the nested recompute backward and therefore has no AccumulateGrad edge in the 
    outer graph. Traversing only the captured forward outputs consequently discovers
   `projection.weight` but misses `checkpointed.weight`, even though backward eventually
    produces a gradient for it.

  ## Solution

  ### Automatically discover hidden checkpoint dependencies

  During CUDA-graph warmup, `CheckpointWithoutOutput` bypasses checkpointing and
  executes its recompute function with autograd enabled. Parameters used by nested
  recompute modules are therefore visible through their `AccumulateGrad` nodes.

  This change:                                                                                                                                             

  - Traverses the warmup autograd graph to discover these parameters.
  - Filters the result to parameters recursively owned by the captured base module.
  - Merges the warmup-discovered parameters with those reachable from the actual
    captured forward outputs.
  - Preserves `base_module.parameters()` ordering so capture and replay use a stable
    positional input layout.
  - Runs at least one internal warmup for training graphs, even when
    `cuda_graph_warmup_steps=0`, because one discovery pass is required for correct
    checkpoint handling.

  This discovery is automatic and does not require checkpoint call sites or
  individual modules to declare additional dependencies.

  ### Capture gradients produced by nested recomputation

  For a parameter hidden behind `CheckpointWithoutOutput`,
  `torch.autograd.grad(..., allow_unused=True)` preserves its requested output slot
  but returns `None` for that slot. Its nested backward instead writes the real
  gradient to `param.grad`.

  Backward capture now transfers that contribution into `param.main_grad`.

  The returned `wgrad` and `param.grad` are handled independently. This is
  important when the same parameter is used both outside and inside checkpoint
  recomputation:

  - `wgrad` represents uses visible to the outer autograd graph.
  - `param.grad` represents uses performed by the nested recompute backward.

  Both are legitimate contributions and must be accumulated exactly once.

  ### Preserve normal DDP bucket readiness during replay

  All discovered parameters are passed as synthetic inputs to
  `_CudagraphReplayNode.apply`. For ordinary DDP parameters, this creates replay
  autograd edges to the existing `AccumulateGrad` nodes where DDP registered its
  hooks.

  For checkpoint parameters whose gradients were already transferred from
  `param.grad` to `param.main_grad`, replay sets
  `grad_added_to_main_grad=True` before the normal DDP hook runs. The hook then:

  - Skips adding the same gradient to `main_grad` a second time.
  - Clears `param.grad`.
  - Still calls `register_grad_ready()`, allowing the DDP bucket to be dispatched.

  GTP parameters retain their existing explicit hook path in
  `_CudagraphReplayNode.backward`.

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### CODE REVIEW

**Comment by JF-D** [comment]
@claude strict-review

**Comment by kvareddy** [comment]
@jiemingz  can you please take a look at this MR?

File: megatron/core/transformer/cuda_graphs.py [hunk 1]
```diff
@@ -1648,6 +1694,12 @@ def replay_graph_capture(self, is_first_microbatch, args, kwargs):
 
         inp_tensors = self.get_tensors(args, kwargs, check_types=False)
         if self.grad_enabled:
+            # Passing parameters as synthetic replay-node inputs creates autograd edges to their
```
Comment (jiemingz) [inline megatron/core/transformer/cuda_graphs.py]: this comment seems not necessary?

File: megatron/core/transformer/cuda_graphs.py [hunk 2]
```diff
@@ -1280,7 +1299,16 @@ def _resolve_input_buffer(ten):
...
             # warmup again as case graph capture mode may execute a different codepath
             _set_warmup_start()
-            for _ in range(self.num_warmup_steps):
+
+            warmup_param_ids = None
+            # The outer autograd graph of CheckpointWithoutOutput has no edges to parameters used
+            # only by its recompute function. Warmup bypasses the checkpoint and runs that function
+            # with autograd, exposing those parameters through AccumulateGrad. Always run at least
+            # one training warmup so they can be included in the captured backward graph inputs.
+            num_internal_warmups = (
```
Comment (jiemingz) [inline megatron/core/transformer/cuda_graphs.py]: can you additionally guard only if the transformer config has recompute set? Ie creating a flag like`needs_recompute_param_discovery`

File: megatron/core/transformer/cuda_graphs.py [hunk 3]
```diff
@@ -1295,6 +1323,8 @@ def clone_ten(ten):
                 if self.grad_enabled:
                     warmup_outputs = self.get_tensors(warmup_outputs)
                     warmup_outputs = tuple(o for o in warmup_outputs if o.requires_grad)
+                    if warmup_param_ids is None:
+                        warmup_param_ids = self.get_connected_param_ids(warmup_outputs)
```
Comment (jiemingz) [inline megatron/core/transformer/cuda_graphs.py]: params are static, why can't we just do:
`self.params_to_backprop =  self.get_connected_param_ids(warmup_outputs)` ?

File: megatron/core/transformer/cuda_graphs.py [hunk 4]
```diff
@@ -1466,9 +1498,23 @@ def create_bwd_graph(self):
             n_act_grads = sum(
                 1 for i in self.fwd_graph_input_surface[: self.num_dgrads] if i.requires_grad
             )
+            self.checkpoint_param_grad_ids = set()
```
Comment (jiemingz) [inline megatron/core/transformer/cuda_graphs.py]: isnt this true broadly? if any function (recompute or not) returns a param.grad we need to accumulate it to main_grad here. The backward func returns `None` grads so all gradients have to be consumed in the graph.

Comment (JF-D) [reply megatron/core/transformer/cuda_graphs.py]: Yes, the tracking set serves a narrower replay purpose and has been renamed to param_ids_to_skip_ddp_grad_add_on_replay. It records only parameters whose param.grad was already transferred to main_grad inside the captured backward graph, but the grad_added_to_main_grad is not set by the DDP hook replay.

**Comment by JF-D** [comment]
@claude strict-review

File: megatron/core/transformer/cuda_graphs.py [hunk 5]
```diff
@@ -1466,9 +1482,17 @@ def create_bwd_graph(self):
             n_act_grads = sum(
                 1 for i in self.fwd_graph_input_surface[: self.num_dgrads] if i.requires_grad
             )
+            self.param_ids_to_skip_ddp_grad_add_on_replay = set()
             for param, wgrad in zip(self.params_to_backprop, grad_inputs[n_act_grads:]):
-                if wgrad is not None and not getattr(param, 'grad_added_to_main_grad', False):
+                grad_already_accumulated = getattr(param, 'grad_added_to_main_grad', False)
+                if wgrad is not None and not grad_already_accumulated:
                     param.main_grad.add_(wgrad)
+                if param.grad is not None and not grad_already_accumulated:
```
Comment (jiemingz) [inline megatron/core/transformer/cuda_graphs.py:1485]: do we even handle cases where `param.grad_added_to_main_grad = False` ? Because the backward replay returns no gradients, meaning all grads need to accumulated at this point. 
`param.grad_added_to_main_grad = False` means it needs to be accumulated by DDP but we return no grads.

Comment (JF-D) [reply megatron/core/transformer/cuda_graphs.py:1485]: grad_added_to_main_grad is default to be False, and it's used to skip the accumulation of DDP hook when set to be true. Here, `grad_added_to_main_grad = False` means the grad hasn't been added to main_grad, so we added them into main grad in the bwd graph. And the grad_added_to_main_grad will be set to true afterwards.

Comment (JF-D) [reply megatron/core/transformer/cuda_graphs.py:1485]: Here https://github.com/NVIDIA/Megatron-LM/pull/6373/changes#diff-a3f9fb509ad46e999a7d8e1b6850c40d2ca0a85344c749096008fefebf51ab08R867-R870 and https://github.com/NVIDIA/Megatron-LM/pull/6373/changes#diff-a3f9fb509ad46e999a7d8e1b6850c40d2ca0a85344c749096008fefebf51ab08R887-R888, we will set the grad_added_to_main_grad to be true.

**Comment by JF-D** [comment]
@claude strict-review

**Review by jiemingz — APPROVED** [review]

**Review by yaoyu-33 — APPROVED** [review]
